Documentation/guides/OneDotNetEmbeddedResources.md

```diff
@@ -0,0 +1,280 @@
+
+There is not a great place where all Android file types would fit
+following this pattern.
+
+In Android Studio, Android libraries are packaged as [`.aar`][aar]
+files. A Xamarin.Android library, `Foo.csproj`, could also generate an
+[`.aar`][aar] file in its `$(OutputPath)`:
+
+    Foo.aar
+        classes.jar
+        res/
+        assets/
+        libs/*.jar
+        jni/[arch]/*.so
+
+`@(AndroidEnvironment)` files are specific to Xamarin.Android, so we
+can make one addition to the file format:
+
+    Foo.aar
+        env/MyEnvironment.txt
+
+So the output of `Foo.csproj` would look like:
+
+    bin/Debug/[targetframework]/
+        Foo.dll
+        Foo.aar
+    bin/Release/[targetframework]/
+        Foo.dll
+        Foo.aar
+
+If you ran the `Pack` target, you would get a `.nupkg` file with:
+
+    lib
+        net6.0-android29\Foo.dll
+        net6.0-android29\Foo.aar
+        net6.0-android30\Foo.dll
+        net6.0-android30\Foo.aar
+
+When consuming the `.nupkg` files, Xamarin.Android will still have to
+extract [`.aar`][aar] files on disk for command-line tooling such as
+`javac`, `d8/r8` and `manifestmerger`. If users want to copy around
+loose build output and consume it, there will only be 1 additional
+[`.aar`][aar] file they will need to copy.
+
+### Java/Kotlin Dependencies
+
+A `Foo.csproj` might include `bar.aar` and `baz.jar` files that are
+Java/Kotlin dependencies.
+
+`.aar` files should sit alongside the .NET assembly:
+
+    lib
+        net6.0-android29\Foo.dll
+        net6.0-android29\Foo.aar
+        net6.0-android29\bar.aar
+        net6.0-android30\Foo.dll
+        net6.0-android30\Foo.aar
+        net6.0-android30\bar.aar
+
+The `baz.jar` file would be included in the above `Foo.aar` file at:
+
+    Foo.aar
+        classes.jar
+        libs/baz.jar
+
+### Native Libraries
+
+Since both `.aar` and `.nupkg` files support native libraries,
+Xamarin.Android should support consuming native libraries from both
+locations. A Xamarin.Android class library, `Foo.csproj` will place
+native libraries in the `Foo.aar` file by default.
+
+### Assembly-Level Attributes
+
+To mark each `.aar` file that is needed for an assembly, a new custom
+attribute is needed:
+
+```csharp
+[assembly: Android.Runtime.Dependency ("Foo.aar")]
+[assembly: Android.Runtime.Dependency ("bar.aar")]
+```
+
+We can emit a build error if any files are missing.
+
+We also need to emit an attribute to mark the assembly as using the
+new system:
+
+```csharp
+[assembly: Android.Runtime.Features (Android.Runtime.Feature.NativeDependencies)]
+```
+
+`Feature.NativeDependencies` would allow us to skip the assembly
+during the `<ResolveLibraryProjectImports/>` MSBuild task. We would
+need to emit this for every Xamarin.Android class library, as new
+libraries might not contain any `[assembly: Dependency]` attributes.
+
+Usage of an enum would allow us to add new things later:
+
+```csharp
+public enum Feature
+{
+    NativeDependencies   = 0x1,
```

Comment:
This feels error-prone.  If I add a new "type" of resource to my NuGet package, I must remember to add a new `Feature` flag, otherwise my new resource won't be included.  This just feels painful, and I'm not sure I understand the actual build time benefit, as we'd have to crack open an assembly to read the attribute *anyway*…

Reply:
Developers won't actually write these attributes, the build adds them automatically.

We could go the route of using a single attribute and a flags-based enum, or create lots of attributes?

In the future, the build will need to know if an assembly has no `EmbeddedResource` at all and skip looking at it in `<ResolveLibraryProjectImports/>`. Future changes could possibly skip an assembly for `<GenerateJavaStubs/>` if we pregenerate `AndroidManifest.xml` or java stubs.

Reply:
> the build adds them automatically

How would the build go about adding them?  Something like the `[assembly:TargetFrameworkAttribute]` done at build time?

Reply:
> In the future, the build will need to know if an assembly has no `EmbeddedResource` at all and skip looking at it in `<ResolveLibraryProjectImports/>`.

Yes, but…. Could we (continue to) differentiate between "initial" builds and "incremental" builds?  In an incremental build scenario, there should be no need to re-process the assembly in the first place, so the custom attributes shouldn't help there.

That leaves the "initial" build scenario, *or* a "I just updated a referenced library project, and now I'm rebuilding the app project" scenario, and *either way* if we have to open the assembly *anyway* (just to check for custom attributes), I'm not sure how much "better" that would be vs. attempting to extract `@(EmbeddedResource)`s when an assembly doesn't have any.

(Some timing data might be interesting there…?)

"*Best*" performance improvements will be achieved by *not touching the assembly at all*.  I don't currently see how this enum helps prevent "touching the assembly".

Reply:
Yeah we could generate a `.cs` file like they do:
```
> cat '.\HelloAndroid\obj\Debug\net5.0-android\.NETCoreApp,Version=v5.0.AssemblyAttributes.cs'
// <autogenerated />
using System;
using System.Reflection;
[assembly: global::System.Runtime.Versioning.TargetFrameworkAttribute(".NETCoreApp,Version=v5.0", FrameworkDisplayName = "")]
```

Reply:
> attempting to extract @(EmbeddedResource)s when an assembly doesn't have any.

Maybe I need to see if this can work without any assembly-level attributes at all.

Reply:
I guess the question is, how do you know which `.aar` files go with an assembly reference? We would just look for `*.aar` in the same directory, maybe that is fine?

Reply:
> how do you know which `.aar` files go with an assembly reference?

Do we need to know that? I'm not sure why that would be immediately needed.  (One idea that came to mind was to restrict the `.aar`/`.jar` files provided to `javac` when building the Java Callable Wrappers, but it would likely be easier to just pass *all* the `.aar`/`.jar` files to all `javac` invocations *anyway*…)

Reply:
I don't know if it is a security concern, but anyone could drop a `.jar` or `.aar` file next to a `.dll` file and it would get used throughout the build.

Reply:
This would probably also work in an application project:
```xml
<None Include="foo.aar" CopyToOutputDirectory="PreserveNewest"/>
```

Reply:
> which is maybe fine?

I don't immediately see why that *wouldn't* be fine…